Kernel/Devices/AHCIController.cpp

```diff
@@ -0,0 +1,166 @@
+/*
+* Copyright (c) 2020, Ben Jilks <benjyjilks@gmail.com>
+* All rights reserved.
+*
+* Redistribution and use in source and binary forms, with or without
+* modification, are permitted provided that the following conditions are met:
+*
+* 1. Redistributions of source code must retain the above copyright notice, this
+*    list of conditions and the following disclaimer.
+*
+* 2. Redistributions in binary form must reproduce the above copyright notice,
+*    this list of conditions and the following disclaimer in the documentation
+*    and/or other materials provided with the distribution.
+*
+* THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+* AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+* IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+* DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+* FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+* DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+* SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+* CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+* OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+* OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*/
+
+#include <Kernel/Devices/AHCIController.h>
+#include <Kernel/Devices/AHCIDiskDevice.h>
+#include <Kernel/Devices/AHCIPort.h>
+#include <Kernel/Devices/NullDevice.h>
+#include <Kernel/VM/AnonymousVMObject.h>
+#include <Kernel/VM/MemoryManager.h>
+#include <LibBareMetal/IO.h>
+#include <LibBareMetal/Memory/PhysicalAddress.h>
+
+namespace Kernel {
+
+#define AHCI_PRIMARY_IRQ 16
+
+#define AHCI_GHC 0x04
+#define AHCI_PI 0x0C
+#define AHCI_VS 0x10
+
+#define PCI_Mass_Storage_Class 0x1
+#define PCI_ATA_Controller_Subclass 0x6
+
+#define READ32_HBA(base, member) *(u32*)(base.get() + member)
+
+void AHCIController::create()
+{
+    PCI::Address pci_address;
+
+    PCI::enumerate_all([&pci_address](const PCI::Address& address, PCI::ID) {
+        int class_id = PCI::get_class(address);
+        int subclass_id = PCI::get_subclass(address);
+
+        if (class_id == PCI_Mass_Storage_Class
+            && subclass_id == PCI_ATA_Controller_Subclass) {
+            klog() << "AHCIController: Controller found!";
+            pci_address = address;
+        }
+    });
+
+    new AHCIController(pci_address);
+}
+
+static AHCIController* s_the;
+
+AHCIController::AHCIController(PCI::Address address)
+    : PCI::Device(address, PCI::get_interrupt_line(address))
+    , m_abar_region(MM.allocate_kernel_region(
+          PhysicalAddress(PCI::get_BAR5(address)),
+          PAGE_ROUND_UP(0x1100), "AHCI ABAR",
```

Comment:
Why do we use the value `0x1100` here?

Reply:
What I'm trying to say is that - if you use a constant number, please define it, using an enum or a macro definition. Otherwise, you can use here the `PCI::get_BAR_Space_Size()` method, that returns the amount of memory being used by the device for BAR5 (but you can check for other BARs if you need to).